src/input/backend_ext.rs

```diff
@@ -0,0 +1,50 @@
+use ::input as libinput;
+use smithay::backend::input;
+use smithay::backend::winit::WinitVirtualDevice;
+use smithay::output::Output;
+
+use crate::backend::Backend;
+use crate::niri::State;
+use crate::protocols::virtual_pointer::VirtualPointer;
+
+pub trait NiriInputBackend: input::InputBackend<Device = Self::NiriDevice> {
+    type NiriDevice: NiriInputDevice;
+}
+impl<T: input::InputBackend> NiriInputBackend for T
+where
+    Self::Device: NiriInputDevice,
+{
+    type NiriDevice = Self::Device;
+}
+
+pub trait NiriInputDevice: input::Device {
+    // FIXME: this should maybe be per-event, not per-device,
+    // but it's not clear that this matters in practice?
+    // it might be more obvious once we implement it for libinput
+    fn output(&self, state: &State) -> Option<Output>;
+}
+
+impl NiriInputDevice for libinput::Device {
+    fn output(&self, _state: &State) -> Option<Output> {
+        // FIXME: Allow specifying the output per-device?
+        None
+    }
+}
+
+impl NiriInputDevice for WinitVirtualDevice {
+    fn output(&self, state: &State) -> Option<Output> {
+        match state.backend {
+            Backend::Winit(ref winit) => Some(winit.single_output().clone()),
```

Comment:
Does this actually do anything? Maybe just replace the whole thing with `None` and remove the `single_output()` function?

Reply:
Basically, it asserts that the mouse position on the winit backend *will* break if/when it ever adds another output. Like, it'd try to scale it relative to the global extents of all monitors instead of just the one. That's obviously like, super wrong, for everything on the winit backend.

---

I guess it might also correctly transform the tablet position? If we return None, we hit that code path:

https://github.com/YaLTeR/niri/blob/f337a272a5c94b657429656608921f930a56cabc/src/input/mod.rs#L290

(the `Transform::Normal` here looks suspicious to me)

---

in general, it just makes sure the more relevant known-output code paths are hit. It feels more correct, if nothing else.